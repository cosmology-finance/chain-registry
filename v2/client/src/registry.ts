import { ChainRegistryChainUtil } from './chain-util';
import { ChainRegistryFetcher, ChainRegistryFetcherOptions } from './fetcher';

export interface ChainRegistryClientOptions
  extends ChainRegistryFetcherOptions {
  chainNames: string[];
  assetListNames?: string[];
  ibcNamePairs?: string[][];
  baseUrl?: string;
}

export class ChainRegistryClient extends ChainRegistryFetcher {
  protected _options: ChainRegistryClientOptions = {
    chainNames: [],
    baseUrl: 'https://raw.githubusercontent.com/chain-registry/chain-registry/main/registries/minimal'
  };

  constructor(options: ChainRegistryClientOptions) {
    const {
      chainNames,
      assetListNames,
      ibcNamePairs,
      baseUrl,
      ...restOptions
    } = options;

    super(restOptions);
    this._options = {
      ...this._options,
      chainNames: chainNames || this._options.chainNames,
      assetListNames: assetListNames || this._options.assetListNames,
      ibcNamePairs: ibcNamePairs || this._options.ibcNamePairs,
      baseUrl: baseUrl || this._options.baseUrl
    };

    this.generateUrls();
  }

  generateUrls() {
    const { chainNames, assetListNames, ibcNamePairs, baseUrl } = this._options;

    const chainUrls = chainNames.map((chain) => {
      return `${baseUrl}/${chain}/chain.json`;
    });

    const assetlistUrls = (assetListNames || chainNames).map((chain) => {
      return `${baseUrl}/${chain}/assetlist.json`;
    });

    let namePairs = ibcNamePairs;
    if (!namePairs) {
      namePairs = [];
      for (let i = 0; i < chainNames.length; i++) {
        for (let j = i + 1; j < chainNames.length; j++) {
          namePairs.push([chainNames[i], chainNames[j]]);
        }
      }
    }
    const ibcUrls = namePairs.map((namePair) => {
      const fileName =
        namePair[0].localeCompare(namePair[1]) <= 0
          ? `${namePair[0]}-${namePair[1]}.json`
          : `${namePair[1]}-${namePair[0]}.json`;
      return `${baseUrl}/_IBC/${fileName}`;
    });

    this.urls = [
      ...new Set([
        ...chainUrls,
        ...assetlistUrls,
        ...ibcUrls,
        ...(this.urls || [])
      ])
    ];
  }

  getChainUtil(chainName: string) {
    const chainInfo = this.getChainInfo(chainName);
    return new ChainRegistryChainUtil({
      chainName: chainName,
      chainInfo: chainInfo
    });
  }
}
