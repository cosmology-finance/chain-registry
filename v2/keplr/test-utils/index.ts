import { AssetList, Chain, IBCInfo } from '@chain-registry/types';

import _assets from '../../../__fixtures__/assets.json';
import _chains from '../../../__fixtures__/chains.json';
import _ibc from '../../../__fixtures__/ibc.json';

export const assets: AssetList[] = _assets as AssetList[];
// TODO update fixtures, they may be missing properties and/or old schema
export const chains: Chain[] = _chains as any;

export const ibc: IBCInfo[] = _ibc as IBCInfo[];
