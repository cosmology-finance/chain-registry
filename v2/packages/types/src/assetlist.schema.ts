export interface Asset {
  deprecated?: boolean;
  description?: string;
  address?: string;
  base: string;
  name: string;
  display: string;
  symbol: string;
  traces?: (IbcTransition | IbcCw20Transition | IbcBridgeTransition | NonIbcTransition)[];
  ibc?: {
    sourceChannel: string;
    dstChannel: string;
    sourceDenom: string;
  };
  images?: {
    png?: string;
    svg?: string;
    theme?: {
      circle?: boolean;
      monochrome?: boolean;
      primaryColorHex?: string;
      backgroundColorHex?: string;
      darkMode?: boolean;
    };
    imageSync?: Pointer;
  }[];
  keywords?: string[];
  socials?: {
    website?: string;
    twitter?: string;
    telegram?: string;
    discord?: string;
    github?: string;
    medium?: string;
    reddit?: string;
  };
  assetType: "sdk.coin" | "cw20" | "erc20" | "ics20" | "snip20" | "snip25" | "bitcoin-like" | "evm-base" | "svm-base" | "substrate" | "unknown" | "sdk.factory" | "bitsong";
  extendedDescription?: string;
  denomUnits: DenomUnit[];
  logoURIs?: {
    png?: string;
    svg?: string;
  };
  coingeckoId?: string;
}
export interface Pointer {
  chainName: string;
  baseDenom?: string;
}
export interface DenomUnit {
  denom: string;
  exponent: number;
  aliases?: string[];
}
export interface IbcTransition {
  type: "ibc";
  counterparty: {
    chainName: string;
    baseDenom: string;
    channelId: string;
  };
  chain: {
    channelId: string;
    // temporarily optional (talk with Jeremy)
    path?: string;
  };
}
export interface IbcCw20Transition {
  type: "ibc-cw20";
  counterparty: {
    port: string;
    chainName: string;
    baseDenom: string;
    channelId: string;
  };
  chain: {
    port: string;
    channelId: string;
    // temporarily optional (talk with Jeremy)
    path?: string;
  };
}
export interface IbcBridgeTransition {
  type: "ibc-bridge";
  counterparty: {
    port?: string;
    chainName: string;
    baseDenom: string;
    channelId: string;
  };
  chain: {
    port?: string;
    channelId: string;
    // temporarily optional (talk with Jeremy)
    path?: string;
  };
  provider: string;
}
export interface NonIbcTransition {
  type: "bridge" | "liquid-stake" | "synthetic" | "wrapped" | "additional-mintage" | "test-mintage" | "legacy-mintage";
  counterparty: {
    contract?: string;
    chainName: string;
    baseDenom: string;
  };
  chain?: {
    contract: string;
  };
  provider: string;
}
export interface AssetList {
  $schema?: string;
  assets: Asset[];
  chainName: string;
}