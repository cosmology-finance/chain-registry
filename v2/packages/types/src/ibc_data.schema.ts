export interface ChainOperatorInfo {
  address?: string;
}
export interface ChainInfo {
  chainName: string;
  clientId: string;
  connectionId: string;
}
export interface ChannelInfo {
  channelId: string;
  portId: string;
  clientId?: string;
  connectionId?: string;
}
export interface IBCData {
  $schema?: string;
  channels: {
    ordering: "ordered" | "unordered";
    version: string;
    description?: string;
    tags?: {
      status?: "live" | "upcoming" | "killed";
      preferred?: boolean;
      dex?: string;
      properties?: string;
    };
    chain1: ChannelInfo;
    chain2: ChannelInfo;
    feeVersion?: string;
  }[];
  operators?: {
    memo: string;
    name: string;
    chain1: ChainOperatorInfo;
    chain2: ChainOperatorInfo;
    discordHandle?: string;
  }[];
  chain1: ChainInfo;
  chain2: ChainInfo;
}