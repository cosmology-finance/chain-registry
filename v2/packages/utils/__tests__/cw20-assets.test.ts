import { getCw20Assets } from '../src';
import { assetLists, ibc as ibcData } from '../test-utils';

it('getCw20Assets', () => {
  const chainName = 'osmosis';
  const list = getCw20Assets(chainName, ibcData, assetLists);

  expect(list).toMatchSnapshot();
});
