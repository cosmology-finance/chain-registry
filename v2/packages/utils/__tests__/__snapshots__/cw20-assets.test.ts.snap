// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCw20Assets 1`] = `
[
  {
    "assets": [
      {
        "address": "juno168ctmpyppk90d34p3jjy658zf5a5l3w8wk35wht6ccqj4mr0yv8s4j5awr",
        "base": "ibc/297C64CC42B5A8D8F82FE2EBE208A6FE8F94B86037FA28C4529A23701C228F7A",
        "coingeckoId": "neta",
        "denomUnits": [
          {
            "aliases": [
              "cw20:juno168ctmpyppk90d34p3jjy658zf5a5l3w8wk35wht6ccqj4mr0yv8s4j5awr",
            ],
            "denom": "ibc/297C64CC42B5A8D8F82FE2EBE208A6FE8F94B86037FA28C4529A23701C228F7A",
            "exponent": 0,
          },
          {
            "denom": "neta",
            "exponent": 6,
          },
        ],
        "description": "The native token cw20 for Neta on Juno Chain",
        "display": "neta",
        "logoURIs": {
          "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/juno/images/neta.png",
          "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/juno/images/neta.svg",
        },
        "name": "Neta",
        "symbol": "NETA",
        "traces": [
          {
            "chain": {
              "channel_id": "channel-169",
              "port": "transfer",
            },
            "counterparty": {
              "base_denom": "cw20:juno168ctmpyppk90d34p3jjy658zf5a5l3w8wk35wht6ccqj4mr0yv8s4j5awr",
              "chain_name": "juno",
              "channel_id": "channel-47",
              "port": "wasm.juno1v4887y83d6g28puzvt8cl0f3cdhd3y6y9mpysnsp3k8krdm7l6jqgm0rkn",
            },
            "type": "ibc-cw20",
          },
        ],
        "typeAsset": "cw20",
      },
      {
        "address": "juno1g2g7ucurum66d42g8k5twk34yegdq8c82858gz0tq2fc75zy7khssgnhjl",
        "base": "ibc/F6B691D5F7126579DDC87357B09D653B47FDCE0A3383FF33C8D8B544FE29A8A6",
        "coingeckoId": "marble",
        "denomUnits": [
          {
            "aliases": [
              "cw20:juno1g2g7ucurum66d42g8k5twk34yegdq8c82858gz0tq2fc75zy7khssgnhjl",
            ],
            "denom": "ibc/F6B691D5F7126579DDC87357B09D653B47FDCE0A3383FF33C8D8B544FE29A8A6",
            "exponent": 0,
          },
          {
            "denom": "marble",
            "exponent": 3,
          },
        ],
        "description": "The native token cw20 for Marble DAO on Juno Chain",
        "display": "marble",
        "logoURIs": {
          "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/juno/images/marble.png",
          "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/juno/images/marble.svg",
        },
        "name": "Marble",
        "symbol": "MARBLE",
        "traces": [
          {
            "chain": {
              "channel_id": "channel-169",
              "port": "transfer",
            },
            "counterparty": {
              "base_denom": "cw20:juno1g2g7ucurum66d42g8k5twk34yegdq8c82858gz0tq2fc75zy7khssgnhjl",
              "chain_name": "juno",
              "channel_id": "channel-47",
              "port": "wasm.juno1v4887y83d6g28puzvt8cl0f3cdhd3y6y9mpysnsp3k8krdm7l6jqgm0rkn",
            },
            "type": "ibc-cw20",
          },
        ],
        "typeAsset": "cw20",
      },
      {
        "address": "juno1re3x67ppxap48ygndmrc7har2cnc7tcxtm9nplcas4v0gc3wnmvs3s807z",
        "base": "ibc/C2A2E9CA95DDD4828B75124B5E27B8401C7D8493BC48353D418CBFC04565899B",
        "coingeckoId": "hope-galaxy",
        "denomUnits": [
          {
            "aliases": [
              "cw20:juno1re3x67ppxap48ygndmrc7har2cnc7tcxtm9nplcas4v0gc3wnmvs3s807z",
            ],
            "denom": "ibc/C2A2E9CA95DDD4828B75124B5E27B8401C7D8493BC48353D418CBFC04565899B",
            "exponent": 0,
          },
          {
            "denom": "hope",
            "exponent": 6,
          },
        ],
        "description": "Hope Galaxy is an NFT collection based on its own native Token $HOPE, a cw20 token on Juno chain.",
        "display": "hope",
        "logoURIs": {
          "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/juno/images/hope.png",
          "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/juno/images/hope.svg",
        },
        "name": "Hope Galaxy",
        "symbol": "HOPE",
        "traces": [
          {
            "chain": {
              "channel_id": "channel-169",
              "port": "transfer",
            },
            "counterparty": {
              "base_denom": "cw20:juno1re3x67ppxap48ygndmrc7har2cnc7tcxtm9nplcas4v0gc3wnmvs3s807z",
              "chain_name": "juno",
              "channel_id": "channel-47",
              "port": "wasm.juno1v4887y83d6g28puzvt8cl0f3cdhd3y6y9mpysnsp3k8krdm7l6jqgm0rkn",
            },
            "type": "ibc-cw20",
          },
        ],
        "typeAsset": "cw20",
      },
      {
        "address": "juno1r4pzw8f9z0sypct5l9j906d47z998ulwvhvqe5xdwgy8wf84583sxwh0pa",
        "base": "ibc/6BDB4C8CCD45033F9604E4B93ED395008A753E01EECD6992E7D1EA23D9D3B788",
        "coingeckoId": "racoon",
        "denomUnits": [
          {
            "aliases": [
              "cw20:juno1r4pzw8f9z0sypct5l9j906d47z998ulwvhvqe5xdwgy8wf84583sxwh0pa",
            ],
            "denom": "ibc/6BDB4C8CCD45033F9604E4B93ED395008A753E01EECD6992E7D1EA23D9D3B788",
            "exponent": 0,
          },
          {
            "denom": "rac",
            "exponent": 6,
          },
        ],
        "description": "Racoon aims to simplify accessibility to AI, NFTs and Gambling on the Cosmos Ecosystem",
        "display": "rac",
        "logoURIs": {
          "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/juno/images/rac.png",
          "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/juno/images/rac.svg",
        },
        "name": "Racoon",
        "symbol": "RAC",
        "traces": [
          {
            "chain": {
              "channel_id": "channel-169",
              "port": "transfer",
            },
            "counterparty": {
              "base_denom": "cw20:juno1r4pzw8f9z0sypct5l9j906d47z998ulwvhvqe5xdwgy8wf84583sxwh0pa",
              "chain_name": "juno",
              "channel_id": "channel-47",
              "port": "wasm.juno1v4887y83d6g28puzvt8cl0f3cdhd3y6y9mpysnsp3k8krdm7l6jqgm0rkn",
            },
            "type": "ibc-cw20",
          },
        ],
        "typeAsset": "cw20",
      },
      {
        "address": "juno1y9rf7ql6ffwkv02hsgd4yruz23pn4w97p75e2slsnkm0mnamhzysvqnxaq",
        "base": "ibc/DB9755CB6FE55192948AE074D18FA815E1429D3D374D5BDA8D89623C6CF235C3",
        "denomUnits": [
          {
            "aliases": [
              "cw20:juno1y9rf7ql6ffwkv02hsgd4yruz23pn4w97p75e2slsnkm0mnamhzysvqnxaq",
            ],
            "denom": "ibc/DB9755CB6FE55192948AE074D18FA815E1429D3D374D5BDA8D89623C6CF235C3",
            "exponent": 0,
          },
          {
            "denom": "block",
            "exponent": 6,
          },
        ],
        "description": "The native token of Marble DEX on Juno Chain",
        "display": "block",
        "logoURIs": {
          "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/juno/images/block.png",
          "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/juno/images/block.svg",
        },
        "name": "Block",
        "symbol": "BLOCK",
        "traces": [
          {
            "chain": {
              "channel_id": "channel-169",
              "port": "transfer",
            },
            "counterparty": {
              "base_denom": "cw20:juno1y9rf7ql6ffwkv02hsgd4yruz23pn4w97p75e2slsnkm0mnamhzysvqnxaq",
              "chain_name": "juno",
              "channel_id": "channel-47",
              "port": "wasm.juno1v4887y83d6g28puzvt8cl0f3cdhd3y6y9mpysnsp3k8krdm7l6jqgm0rkn",
            },
            "type": "ibc-cw20",
          },
        ],
        "typeAsset": "cw20",
      },
      {
        "address": "juno1tdjwrqmnztn2j3sj2ln9xnyps5hs48q3ddwjrz7jpv6mskappjys5czd49",
        "base": "ibc/52E12CF5CA2BB903D84F5298B4BFD725D66CAB95E09AA4FC75B2904CA5485FEB",
        "denomUnits": [
          {
            "aliases": [
              "cw20:juno1tdjwrqmnztn2j3sj2ln9xnyps5hs48q3ddwjrz7jpv6mskappjys5czd49",
            ],
            "denom": "ibc/52E12CF5CA2BB903D84F5298B4BFD725D66CAB95E09AA4FC75B2904CA5485FEB",
            "exponent": 0,
          },
        ],
        "description": "The DAO token to build consensus among Hong Kong People",
        "display": "dhk",
        "logoURIs": {
          "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/juno/images/dhk.png",
          "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/juno/images/dhk.svg",
        },
        "name": "DHK",
        "symbol": "DHK",
        "traces": [
          {
            "chain": {
              "channel_id": "channel-169",
              "port": "transfer",
            },
            "counterparty": {
              "base_denom": "cw20:juno1tdjwrqmnztn2j3sj2ln9xnyps5hs48q3ddwjrz7jpv6mskappjys5czd49",
              "chain_name": "juno",
              "channel_id": "channel-47",
              "port": "wasm.juno1v4887y83d6g28puzvt8cl0f3cdhd3y6y9mpysnsp3k8krdm7l6jqgm0rkn",
            },
            "type": "ibc-cw20",
          },
        ],
        "typeAsset": "cw20",
      },
      {
        "address": "juno15u3dt79t6sxxa3x3kpkhzsy56edaa5a66wvt3kxmukqjz2sx0hes5sn38g",
        "base": "ibc/00B6E60AD3D65CBEF5579AC8AF609527C0B57535B6E32D96C80A735344FD9DCC",
        "coingeckoId": "junoswap-raw-dao",
        "denomUnits": [
          {
            "aliases": [
              "cw20:juno15u3dt79t6sxxa3x3kpkhzsy56edaa5a66wvt3kxmukqjz2sx0hes5sn38g",
            ],
            "denom": "ibc/00B6E60AD3D65CBEF5579AC8AF609527C0B57535B6E32D96C80A735344FD9DCC",
            "exponent": 0,
          },
          {
            "denom": "raw",
            "exponent": 6,
          },
        ],
        "description": "Token governance for Junoswap",
        "display": "raw",
        "logoURIs": {
          "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/juno/images/raw.png",
        },
        "name": "JunoSwap",
        "symbol": "RAW",
        "traces": [
          {
            "chain": {
              "channel_id": "channel-169",
              "port": "transfer",
            },
            "counterparty": {
              "base_denom": "cw20:juno15u3dt79t6sxxa3x3kpkhzsy56edaa5a66wvt3kxmukqjz2sx0hes5sn38g",
              "chain_name": "juno",
              "channel_id": "channel-47",
              "port": "wasm.juno1v4887y83d6g28puzvt8cl0f3cdhd3y6y9mpysnsp3k8krdm7l6jqgm0rkn",
            },
            "type": "ibc-cw20",
          },
        ],
        "typeAsset": "cw20",
      },
      {
        "address": "juno17wzaxtfdw5em7lc94yed4ylgjme63eh73lm3lutp2rhcxttyvpwsypjm4w",
        "base": "ibc/AA1C80225BCA7B32ED1FC6ABF8B8E899BEB48ECDB4B417FD69873C6D715F97E7",
        "denomUnits": [
          {
            "aliases": [
              "cw20:juno17wzaxtfdw5em7lc94yed4ylgjme63eh73lm3lutp2rhcxttyvpwsypjm4w",
            ],
            "denom": "ibc/AA1C80225BCA7B32ED1FC6ABF8B8E899BEB48ECDB4B417FD69873C6D715F97E7",
            "exponent": 0,
          },
          {
            "denom": "asvt",
            "exponent": 6,
          },
        ],
        "description": "Profit sharing token for Another.Software validator. Hold and receive dividends from Another.Software validator commissions!",
        "display": "asvt",
        "logoURIs": {
          "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/juno/images/asvt.png",
        },
        "name": "Another.Software Validator Token",
        "symbol": "ASVT",
        "traces": [
          {
            "chain": {
              "channel_id": "channel-169",
              "port": "transfer",
            },
            "counterparty": {
              "base_denom": "cw20:juno17wzaxtfdw5em7lc94yed4ylgjme63eh73lm3lutp2rhcxttyvpwsypjm4w",
              "chain_name": "juno",
              "channel_id": "channel-47",
              "port": "wasm.juno1v4887y83d6g28puzvt8cl0f3cdhd3y6y9mpysnsp3k8krdm7l6jqgm0rkn",
            },
            "type": "ibc-cw20",
          },
        ],
        "typeAsset": "cw20",
      },
      {
        "address": "juno1ur4jx0sxchdevahep7fwq28yk4tqsrhshdtylz46yka3uf6kky5qllqp4k",
        "base": "ibc/D53609853919635406B7F9B83BADE77BF480F7030215B49E932A1EF0A193E49A",
        "denomUnits": [
          {
            "aliases": [
              "cw20:juno1ur4jx0sxchdevahep7fwq28yk4tqsrhshdtylz46yka3uf6kky5qllqp4k",
            ],
            "denom": "ibc/D53609853919635406B7F9B83BADE77BF480F7030215B49E932A1EF0A193E49A",
            "exponent": 0,
          },
          {
            "denom": "hns",
            "exponent": 6,
          },
        ],
        "description": "IBC HNS is HNS, coin of Handshake, decentralized root namesystem, but wrapped to cosmos for IBC support by Another.Software through Juno netowrk.",
        "display": "hns",
        "logoURIs": {
          "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/juno/images/hns.svg",
        },
        "name": "IBC HNS (Handshake)",
        "symbol": "HNS",
        "traces": [
          {
            "chain": {
              "channel_id": "channel-169",
              "port": "transfer",
            },
            "counterparty": {
              "base_denom": "cw20:juno1ur4jx0sxchdevahep7fwq28yk4tqsrhshdtylz46yka3uf6kky5qllqp4k",
              "chain_name": "juno",
              "channel_id": "channel-47",
              "port": "wasm.juno1v4887y83d6g28puzvt8cl0f3cdhd3y6y9mpysnsp3k8krdm7l6jqgm0rkn",
            },
            "type": "ibc-cw20",
          },
        ],
        "typeAsset": "cw20",
      },
      {
        "address": "juno1n7n7d5088qlzlj37e9mgmkhx6dfgtvt02hqxq66lcap4dxnzdhwqfmgng3",
        "base": "ibc/0CB9DB3441D0D50F35699DEE22B9C965487E83FB2D9F483D1CC5CA34E856C484",
        "denomUnits": [
          {
            "aliases": [
              "cw20:juno1n7n7d5088qlzlj37e9mgmkhx6dfgtvt02hqxq66lcap4dxnzdhwqfmgng3",
            ],
            "denom": "ibc/0CB9DB3441D0D50F35699DEE22B9C965487E83FB2D9F483D1CC5CA34E856C484",
            "exponent": 0,
          },
          {
            "denom": "joe",
            "exponent": 6,
          },
        ],
        "description": "DAO dedicated to building tools on the Juno Network",
        "display": "joe",
        "logoURIs": {
          "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/juno/images/joe.png",
        },
        "name": "JoeDAO",
        "symbol": "JOE",
        "traces": [
          {
            "chain": {
              "channel_id": "channel-169",
              "port": "transfer",
            },
            "counterparty": {
              "base_denom": "cw20:juno1n7n7d5088qlzlj37e9mgmkhx6dfgtvt02hqxq66lcap4dxnzdhwqfmgng3",
              "chain_name": "juno",
              "channel_id": "channel-47",
              "port": "wasm.juno1v4887y83d6g28puzvt8cl0f3cdhd3y6y9mpysnsp3k8krdm7l6jqgm0rkn",
            },
            "type": "ibc-cw20",
          },
        ],
        "typeAsset": "cw20",
      },
      {
        "address": "juno1sfwye65qxcfsc837gu5qcprcz7w49gkv3wnat04764ld76hy3arqs779tr",
        "base": "ibc/500F3E928AB5F2E9ACB113F4F879A44BDB42CCDCDC51254222D742A240A526F8",
        "denomUnits": [
          {
            "aliases": [
              "cw20:juno1sfwye65qxcfsc837gu5qcprcz7w49gkv3wnat04764ld76hy3arqs779tr",
            ],
            "denom": "ibc/500F3E928AB5F2E9ACB113F4F879A44BDB42CCDCDC51254222D742A240A526F8",
            "exponent": 0,
          },
          {
            "denom": "dla",
            "exponent": 6,
          },
        ],
        "description": "Governance Token for Digital Land Acquisition DAO",
        "display": "dla",
        "logoURIs": {
          "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/juno/images/dla.png",
          "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/juno/images/dla.svg",
        },
        "name": "Digital Land Acquisition DAO",
        "symbol": "DLA",
        "traces": [
          {
            "chain": {
              "channel_id": "channel-169",
              "port": "transfer",
            },
            "counterparty": {
              "base_denom": "cw20:juno1sfwye65qxcfsc837gu5qcprcz7w49gkv3wnat04764ld76hy3arqs779tr",
              "chain_name": "juno",
              "channel_id": "channel-47",
              "port": "wasm.juno1v4887y83d6g28puzvt8cl0f3cdhd3y6y9mpysnsp3k8krdm7l6jqgm0rkn",
            },
            "type": "ibc-cw20",
          },
        ],
        "typeAsset": "cw20",
      },
      {
        "address": "juno1j0a9ymgngasfn3l5me8qpd53l5zlm9wurfdk7r65s5mg6tkxal3qpgf5se",
        "base": "ibc/52C57FCA7D6854AA178E7A183DDBE4EF322B904B1D719FC485F6FFBC1F72A19E",
        "denomUnits": [
          {
            "aliases": [
              "cw20:juno1j0a9ymgngasfn3l5me8qpd53l5zlm9wurfdk7r65s5mg6tkxal3qpgf5se",
            ],
            "denom": "ibc/52C57FCA7D6854AA178E7A183DDBE4EF322B904B1D719FC485F6FFBC1F72A19E",
            "exponent": 0,
          },
          {
            "denom": "glto",
            "exponent": 6,
          },
        ],
        "description": "DeFi gaming platform built on Juno",
        "display": "glto",
        "logoURIs": {
          "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/juno/images/glto.png",
          "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/juno/images/glto.svg",
        },
        "name": "Gelotto",
        "symbol": "GLTO",
        "traces": [
          {
            "chain": {
              "channel_id": "channel-169",
              "port": "transfer",
            },
            "counterparty": {
              "base_denom": "cw20:juno1j0a9ymgngasfn3l5me8qpd53l5zlm9wurfdk7r65s5mg6tkxal3qpgf5se",
              "chain_name": "juno",
              "channel_id": "channel-47",
              "port": "wasm.juno1v4887y83d6g28puzvt8cl0f3cdhd3y6y9mpysnsp3k8krdm7l6jqgm0rkn",
            },
            "type": "ibc-cw20",
          },
        ],
        "typeAsset": "cw20",
      },
      {
        "address": "juno1gz8cf86zr4vw9cjcyyv432vgdaecvr9n254d3uwwkx9rermekddsxzageh",
        "base": "ibc/7C781B4C2082CD62129A972D47486D78EC17155C299270E3C89348EA026BEAF8",
        "denomUnits": [
          {
            "aliases": [
              "cw20:juno1gz8cf86zr4vw9cjcyyv432vgdaecvr9n254d3uwwkx9rermekddsxzageh",
            ],
            "denom": "ibc/7C781B4C2082CD62129A972D47486D78EC17155C299270E3C89348EA026BEAF8",
            "exponent": 0,
          },
          {
            "denom": "gkey",
            "exponent": 6,
          },
        ],
        "description": "Gelotto Year 1 Grand Prize Token",
        "display": "gkey",
        "logoURIs": {
          "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/juno/images/gkey.png",
          "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/juno/images/gkey.svg",
        },
        "name": "GKey",
        "symbol": "GKEY",
        "traces": [
          {
            "chain": {
              "channel_id": "channel-169",
              "port": "transfer",
            },
            "counterparty": {
              "base_denom": "cw20:juno1gz8cf86zr4vw9cjcyyv432vgdaecvr9n254d3uwwkx9rermekddsxzageh",
              "chain_name": "juno",
              "channel_id": "channel-47",
              "port": "wasm.juno1v4887y83d6g28puzvt8cl0f3cdhd3y6y9mpysnsp3k8krdm7l6jqgm0rkn",
            },
            "type": "ibc-cw20",
          },
        ],
        "typeAsset": "cw20",
      },
      {
        "address": "juno1rdw3gumdz7zyjn2pev9ugxs765xlv6vtv6g3jt2lqw580zrchvjs66daca",
        "base": "ibc/36ABCB2BD7197D778A8EFD4F97E5721CA4CF87F7D39EDD76EC238F2219B7A508",
        "denomUnits": [
          {
            "aliases": [
              "cw20:juno1rdw3gumdz7zyjn2pev9ugxs765xlv6vtv6g3jt2lqw580zrchvjs66daca",
            ],
            "denom": "ibc/36ABCB2BD7197D778A8EFD4F97E5721CA4CF87F7D39EDD76EC238F2219B7A508",
            "exponent": 0,
          },
          {
            "denom": "blkh",
            "exponent": 4,
          },
        ],
        "description": "The native token cw20 for BlackHole on Juno Chain",
        "display": "blkh",
        "logoURIs": {
          "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/juno/images/blkh.png",
          "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/juno/images/blkh.svg",
        },
        "name": "Blackhole",
        "symbol": "BLKH",
        "traces": [
          {
            "chain": {
              "channel_id": "channel-169",
              "port": "transfer",
            },
            "counterparty": {
              "base_denom": "cw20:juno1rdw3gumdz7zyjn2pev9ugxs765xlv6vtv6g3jt2lqw580zrchvjs66daca",
              "chain_name": "juno",
              "channel_id": "channel-47",
              "port": "wasm.juno1v4887y83d6g28puzvt8cl0f3cdhd3y6y9mpysnsp3k8krdm7l6jqgm0rkn",
            },
            "type": "ibc-cw20",
          },
        ],
        "typeAsset": "cw20",
      },
      {
        "address": "ki1dt3lk455ed360pna38fkhqn0p8y44qndsr77qu73ghyaz2zv4whq83mwdy",
        "base": "ibc/AD185F62399F770CCCE8A36A180A77879FF6C26A0398BD3D2A74E087B0BFA121",
        "coingeckoId": "lvn",
        "denomUnits": [
          {
            "aliases": [
              "cw20:ki1dt3lk455ed360pna38fkhqn0p8y44qndsr77qu73ghyaz2zv4whq83mwdy",
            ],
            "denom": "ibc/AD185F62399F770CCCE8A36A180A77879FF6C26A0398BD3D2A74E087B0BFA121",
            "exponent": 0,
          },
          {
            "denom": "lvn",
            "exponent": 6,
          },
        ],
        "description": "ELEVENPARIS loyalty token on KiChain",
        "display": "lvn",
        "logoURIs": {
          "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/kichain/images/lvn.png",
        },
        "name": "LVN",
        "symbol": "LVN",
        "traces": [
          {
            "chain": {
              "channel_id": "channel-261",
              "port": "transfer",
            },
            "counterparty": {
              "base_denom": "cw20:ki1dt3lk455ed360pna38fkhqn0p8y44qndsr77qu73ghyaz2zv4whq83mwdy",
              "chain_name": "kichain",
              "channel_id": "channel-18",
              "port": "wasm.ki1hzz0s0ucrhdp6tue2lxk3c03nj6f60qy463we7lgx0wudd72ctmsd9kgha",
            },
            "type": "ibc-cw20",
          },
        ],
        "typeAsset": "cw20",
      },
    ],
    "chainName": "osmosis",
  },
]
`;
